tofhir-server/src/main/scala/io/tofhir/server/util/CsvUtil.scala

```diff
@@ -0,0 +1,182 @@
+package io.tofhir.server.util
+
+import akka.stream.scaladsl.{Concat, FileIO, Flow, Framing, Sink, Source}
+import akka.util.ByteString
+import com.opencsv.CSVParserBuilder
+import io.tofhir.engine.Execution.actorSystem
+import io.tofhir.engine.Execution.actorSystem.dispatcher
+import io.tofhir.server.common.model.InternalError
+
+import java.io.File
+import java.nio.file.StandardOpenOption
+import scala.concurrent.Future
+
+object CsvUtil {
+
+  /**
+   * Flow to remove carriage returns from a ByteString
+   */
+  private val removeCarriageReturns: Flow[ByteString, ByteString, _] = Flow[ByteString].map { line =>
+    // Remove \r characters from each line
+    val lineStr = line.utf8String
+    val modifiedLineStr = lineStr.replaceAll("\r", "")
+    ByteString(modifiedLineStr)
+  }
+
+  /**
+   * Get the total number of rows in a file
+   * @param file file to count the rows
+   * @return Future[Long] total number of rows in the file
+   */
+  private def getTotalRows(file: File): Future[Long] = {
+    FileIO.fromPath(file.toPath)
+      .via(Framing.delimiter(ByteString("\n"), maximumFrameLength = 1024, allowTruncation = true))
+      // filter out empty lines
+      .filterNot(_.isEmpty)
+      .drop(1)
+      .runFold(0L)((count, _) => count + 1)
+  }
+
+  /**
+   * Overwrite headers string to the first line of the CSV file
+   * Adjust the rows to match the new headers if necessary e.g.
+   * - if a column removed, remove the column from the rows
+   * - if a new column name recognized, add the column to the rows with a default value (<column_name>)
+   * - order of the columns are preserved according to the newHHeaders
+   * @param file CSV file
+   * @param newHeaders Headers to write to the file
+   * @return
+   */
+  def writeCsvHeaders(file: File, newHeaders: Seq[String]): Future[Unit] = {
+    // Create a CSVParser to handle double quotes
+    val parser = new CSVParserBuilder().withSeparator(',').withQuoteChar('"').build()
+
+    // Read the existing CSV file into a list of list where each map is a row
+    val existingContentFuture = FileIO.fromPath(file.toPath)
+      .via(Framing.delimiter(ByteString("\n"), maximumFrameLength = 1024, allowTruncation = true))
+      .map(_.utf8String)
+      .filterNot(_.isEmpty)
+      .runWith(Sink.seq)
+      .map { lines =>
+        if (lines.nonEmpty) {
+          val oldHeaders = parser.parseLine(lines.head.trim)
+          lines.tail.map { line =>
+            oldHeaders.zip(parser.parseLine(line.trim)).toSeq
+          }
+        } else {
+          List.empty
+        }
+      }
+
+    // here existingContent looks like
```

Comment:
Improving the examples here to clarify what oldHeaders include etc. would be fine.